### PR TITLE
feat: `SchemaAdapterStream` may create virtual columns

### PR DESCRIPTION
**Pull request by crepererum**
For chunk order handling in #6098.

This is somewhat similar to what the DataFusion `ParquetExec` is doing.

### CODE REVIEW

File: iox_query/src/provider/adapter.rs [hunk 1]
```diff
@@ -322,23 +359,32 @@ mod tests {
...
-            "| bar |   | 5 |   | 2 |",
-            "| baz |   | 6 |   | 3 |",
-            "+-----+---+---+---+---+",
+            "+-----+---+---+---+-----+---+",
+            "| c   | e | b | d | f   | a |",
+            "+-----+---+---+---+-----+---+",
+            "| foo |   | 4 |   | xxx | 1 |",
+            "| bar |   | 5 |   | xxx | 2 |",
+            "| baz |   | 6 |   | xxx | 3 |",
+            "+-----+---+---+---+-----+---+",
```
Comment (wolffcm) [inline iox_query/src/provider/adapter.rs:387]: So if the virtual column is both in the input stream and declared to be a virtual column, then the real input column wins and is in the output.

Based on my (admittedly not deep) understanding from your notes in #6098 it seems like maybe this should be an error instead? Or is it possible that the schema adapter is applied twice, and you want the adapter to be idempotent?

**Review by wolffcm — APPROVED** [review]
This seems straightforward, but I had a question for my own benefit.

File: iox_query/src/provider/adapter.rs [hunk 2]
```diff
@@ -105,6 +113,7 @@ impl SchemaAdapterStream {
     pub(crate) fn try_new(
```
Comment (alamb) [inline iox_query/src/provider/adapter.rs:113]: Can you please document what virtual columns / constant columns are in this docstring? I think it is a map of `column_name` to the constant value.

File: iox_query/src/provider/adapter.rs (hunk 1 above)
Comment (alamb) [reply iox_query/src/provider/adapter.rs:387]: I suggest the schema adapter should throw an error if the virtual column tries to overrride an actual column. It seems like silently picking one or the other might potentially mask bugs, and make debugging them challenging

But like @wolffcm  I may not appreciate the entire subtlety

**Review by alamb — APPROVED** [review]

File: iox_query/src/provider/adapter.rs (hunk 2 above)
Comment (crepererum) [reply iox_query/src/provider/adapter.rs:113]: done

File: iox_query/src/provider/adapter.rs (hunk 1 above)
Comment (crepererum) [reply iox_query/src/provider/adapter.rs:387]: Good call. I've added an error. While doing so I've also found that the schema check for NULL columns was pointless (it checked the value that it just copied) but that it missed to check if field is actually nullable. I've fixed and tested that as well.

**Pull request merged**